diff --git a/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/SQP.cpython-39.pyc b/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/SQP.cpython-39.pyc
index 2388fb5..ac3f777 100644
Binary files a/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/SQP.cpython-39.pyc and b/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/SQP.cpython-39.pyc differ
diff --git a/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/cost.cpython-39.pyc b/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/cost.cpython-39.pyc
index 84c8ab6..9cd58f1 100644
Binary files a/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/cost.cpython-39.pyc and b/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/cost.cpython-39.pyc differ
diff --git a/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/dynamics.cpython-39.pyc b/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/dynamics.cpython-39.pyc
index 409869e..b7d98ac 100644
Binary files a/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/dynamics.cpython-39.pyc and b/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/__pycache__/dynamics.cpython-39.pyc differ
diff --git a/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/main.py b/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/main.py
index 56e5cee..ac5c496 100644
--- a/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/main.py
+++ b/Coding/OPTCON/03_SQP_equilibria/03_SQP_equilibria/main.py
@@ -187,6 +187,7 @@ hh_end = ff_trim(zz_tr, uu_tr)[0]
 
 print(ll_end, hh_end)
 print(zz_tr, xx_eq_des)
+print(uu_tr, uu_eq_des)
 
 
 
diff --git a/Coding/OPTCON/Exercise_3.ipynb b/Coding/OPTCON/Exercise_3.ipynb
index 6abdcc2..b494ad3 100644
--- a/Coding/OPTCON/Exercise_3.ipynb
+++ b/Coding/OPTCON/Exercise_3.ipynb
@@ -2586,6 +2586,37 @@
     "\n",
     "plt.show()"
    ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "(2, 4)\n"
+     ]
+    }
+   ],
+   "source": [
+    "import numpy as np\n",
+    "tensor = np.random.rand(6,2,4)\n",
+    "vector = np.random.rand(6)\n",
+    "\n",
+    "matric = np.tensordot(tensor, vector, axes = ((0),(0)))\n",
+    "\n",
+    "\n",
+    "print(matric.shape)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
   }
  ],
  "metadata": {
@@ -2604,7 +2635,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.9.13"
+   "version": "3.9.13 (tags/v3.9.13:6de2ca5, May 17 2022, 16:36:42) [MSC v.1929 64 bit (AMD64)]"
   },
   "orig_nbformat": 4,
   "vscode": {
diff --git a/Coding/OPTCON/homework/__pycache__/utils.cpython-39.pyc b/Coding/OPTCON/homework/__pycache__/utils.cpython-39.pyc
index f1ddced..a85b208 100644
Binary files a/Coding/OPTCON/homework/__pycache__/utils.cpython-39.pyc and b/Coding/OPTCON/homework/__pycache__/utils.cpython-39.pyc differ
diff --git a/Coding/OPTCON/homework/main_gradient_optcon_method.py b/Coding/OPTCON/homework/main_gradient_optcon_method.py
index e7c230a..19baab4 100644
--- a/Coding/OPTCON/homework/main_gradient_optcon_method.py
+++ b/Coding/OPTCON/homework/main_gradient_optcon_method.py
@@ -123,7 +123,7 @@ for kk in range(max_iters-1):
   # Descent direction calculation
   ##################################
 
-  # TODO
+  # TODOa
   lmbd_temp = cst.termcost(xx[:,-1,kk], xx_ref[:,-1])[1]
   lmbd[:,-1,kk] = lmbd_temp.squeeze()
 
